File: packages/contracts-java/src/main/java/dev/anvilkit/contracts/SignatureConformance.java
package dev.anvilkit.contracts;

import tools.jackson.databind.JsonNode;
import tools.jackson.databind.json.JsonMapper;
import tools.jackson.databind.node.ArrayNode;
import tools.jackson.databind.node.ObjectNode;

import java.io.ByteArrayOutputStream;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.security.MessageDigest;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Base64;
import java.util.Comparator;
import java.util.HexFormat;
import java.util.List;

/** Emits M4 conformance-result v1 for the shared DSSE/JWS primitive corpus. */
public final class SignatureConformance {
    private static final JsonMapper MAPPER = JsonMapper.builder().build();

    private SignatureConformance() {}

    private static String digest(byte[] raw) throws Exception {
        return "sha256:" + HexFormat.of().formatHex(MessageDigest.getInstance("SHA-256").digest(raw));
    }

    private static byte[] decode(String value) {
        return Base64.getUrlDecoder().decode(value);
    }

    public static ObjectNode generate(Path repositoryRoot) throws Exception {
        if (Runtime.version().feature() != 17) throw new IllegalStateException("expected JDK 17, got " + Runtime.version());
        repositoryRoot = repositoryRoot.toAbsolutePath().normalize();
        Path corpusPath = repositoryRoot.resolve("contracts/governance/m3/signature-cases.json");
        byte[] corpusBytes = Files.readAllBytes(corpusPath);
        JsonNode corpus = MAPPER.readTree(corpusBytes);
        if (corpus.get("corpusVersion").asInt() != 1 || corpus.get("cases").size() != 6) {
            throw new IllegalStateException("invalid signature corpus");
        }
        byte[] publicKey = decode(corpus.at("/key/publicKeyBase64Url").asString());
        byte[] privateSeed = decode(corpus.at("/key/privateSeedBase64Url").asString());
        byte[] dssePayload = decode(corpus.at("/dsse/payloadBase64Url").asString());
        byte[] dsseSignature = decode(corpus.at("/dsse/signatureBase64Url").asString());
        byte[] jwsSignature = decode(corpus.at("/jws/signatureBase64Url").asString());
        List<ObjectNode> cases = new ArrayList<>();
        for (JsonNode vector : corpus.get("cases")) {
            String id = vector.get("id").asString();
            byte[] message;
            byte[] expectedSignature;
            if (vector.get("profile").asString().equals("dsse")) {
                message = ContractSignature.dssePreAuthEncoding(corpus.at("/dsse/payloadType").asString(), dssePayload);
                expectedSignature = dsseSignature;
            } else if (vector.get("profile").asString().equals("jws")) {
                message = (corpus.at("/jws/protectedBase64Url").asString() + "." + corpus.at("/jws/payloadBase64Url").asString())
                        .getBytes(StandardCharsets.US_ASCII);
                expectedSignature = jwsSignature;
            } else throw new IllegalStateException(id + ": unsupported profile");
            message = message.clone();
            if (vector.get("mutation").asString().equals("message-last-byte")) message[message.length - 1] ^= 1;
            byte[] candidateSignature = expectedSignature.clone();
            if (vector.get("operation").asString().equals("sign-and-verify")) {
                candidateSignature = ContractSignature.sign(privateSeed, message);
                if (!Arrays.equals(candidateSignature, expectedSignature)) {
                    throw new IllegalStateException(id + ": deterministic signature differs from corpus");
                }
            }
            if (vector.get("mutation").asString().equals("signature-first-byte")) candidateSignature[0] ^= 1;
            boolean verified = ContractSignature.verify(publicKey, message, candidateSignature);
            if (verified != vector.get("expectedVerified").asBoolean()) {
                throw new IllegalStateException(id + ": verification differs from corpus");
            }
            ObjectNode item = MAPPER.createObjectNode();
            item.put("caseId", id);
            item.put("inputDigest", digest(message));
            item.put("inputBytes", message.length);
            item.put("parseOutcome", "accepted");
            item.put("valid", verified);
            ArrayNode findings = item.putArray("findings");
            if (!verified) {
                ObjectNode finding = findings.addObject();
                finding.put("code", "SIGNATURE_INVALID");
                finding.put("instancePath", "/signature");
                finding.put("schemaPath", "/profile/ed25519");
            }
            item.putObject("canonicalization").put("status", "not-applicable");
            item.putNull("componentDigest");
            item.putNull("rootBomDigest");
            ObjectNode signature = item.putObject("signature");
            signature.put("status", verified ? "verified" : "rejected");
            if (!verified) signature.put("code", "SIGNATURE_INVALID");
            cases.add(item);
        }
        cases.sort(Comparator.comparing(item -> item.get("caseId").asString()));
        ByteArrayOutputStream adapterBytes = new ByteArrayOutputStream();
        for (String name : List.of("ContractSignature.java", "SignatureConformance.java")) {
            adapterBytes.write(Files.readAllBytes(repositoryRoot.resolve("packages/contracts-java/src/main/java/dev/anvilkit/contracts").resolve(name)));
        }
        ObjectNode output = MAPPER.createObjectNode();
        output.put("resultVersion", 1);
        output.put("fixtureManifestDigest", digest(corpusBytes));
        output.put("language", "java");
        ObjectNode implementation = output.putObject("implementation");
        implementation.put("adapterId", "anvilkit-java-signature-native");
        implementation.put("adapterVersion", "0.1.0");
        implementation.put("runtime", "jdk");
        implementation.put("runtimeVersion", "17");
        implementation.put("adapterDigest", digest(adapterBytes.toByteArray()));
        ArrayNode outputCases = output.putArray("cases");
        for (ObjectNode item : cases) outputCases.add(item);
        return output;
    }

    public static void main(String[] args) throws Exception {
        Path repositoryRoot = Path.of("../..");
        int iterations = 1;
        for (int index = 0; index + 1 < args.length; index++) {
            if (args[index].equals("--repository-root")) repositoryRoot = Path.of(args[index + 1]);
            if (args[index].equals("--iterations")) iterations = Integer.parseInt(args[index + 1]);
        }
        if (iterations < 1) throw new IllegalArgumentException("--iterations must be positive");
        ObjectNode result = null;
        for (int index = 0; index < iterations; index++) result = generate(repositoryRoot);
        System.out.println(MAPPER.writerWithDefaultPrettyPrinter().writeValueAsString(result));
    }
}

File: packages/contracts-python/anvilkit_contracts/signature_conformance.py
"""M4 language-neutral results for the shared DSSE/JWS primitive corpus."""

from __future__ import annotations

import argparse
import base64
import hashlib
import importlib.metadata
import json
import platform
from pathlib import Path
from typing import Any

from .signature import dsse_pre_auth_encoding, sign_ed25519, verify_ed25519


def _decode(value: str) -> bytes:
    return base64.urlsafe_b64decode(value + "=" * (-len(value) % 4))


def _digest(raw: bytes) -> str:
    return f"sha256:{hashlib.sha256(raw).hexdigest()}"


def generate(repository_root: Path) -> dict[str, Any]:
    if platform.python_version() != "3.12.7":
        raise RuntimeError(f"expected Python 3.12.7, got {platform.python_version()}")
    version = importlib.metadata.version("cryptography")
    if version != "49.0.0":
        raise RuntimeError(f"expected cryptography 49.0.0, got {version}")
    repository_root = repository_root.resolve()
    corpus_path = repository_root / "contracts/governance/m3/signature-cases.json"
    corpus_bytes = corpus_path.read_bytes()
    corpus = json.loads(corpus_bytes)
    if corpus["corpusVersion"] != 1 or len(corpus["cases"]) != 6:
        raise ValueError("invalid signature corpus")
    public_key = _decode(corpus["key"]["publicKeyBase64Url"])
    private_seed = _decode(corpus["key"]["privateSeedBase64Url"])
    dsse_payload = _decode(corpus["dsse"]["payloadBase64Url"])
    dsse_signature = _decode(corpus["dsse"]["signatureBase64Url"])
    jws_signature = _decode(corpus["jws"]["signatureBase64Url"])
    cases: list[dict[str, Any]] = []
    for vector in corpus["cases"]:
        if vector["profile"] == "dsse":
            original_message = dsse_pre_auth_encoding(corpus["dsse"]["payloadType"], dsse_payload)
            expected_signature = dsse_signature
        elif vector["profile"] == "jws":
            original_message = (
                corpus["jws"]["protectedBase64Url"] + "." + corpus["jws"]["payloadBase64Url"]
            ).encode("ascii")
            expected_signature = jws_signature
        else:
            raise ValueError(f"{vector['id']}: unsupported profile")
        message = bytearray(original_message)
        if vector["mutation"] == "message-last-byte":
            message[-1] ^= 1
        candidate_signature = bytearray(expected_signature)
        if vector["operation"] == "sign-and-verify":
            candidate_signature = bytearray(sign_ed25519(private_seed, bytes(message)))
            if candidate_signature != expected_signature:
                raise ValueError(f"{vector['id']}: deterministic signature differs from corpus")
        if vector["mutation"] == "signature-first-byte":
            candidate_signature[0] ^= 1
        verified = verify_ed25519(public_key, bytes(message), bytes(candidate_signature))
        if verified != vector["expectedVerified"]:
            raise ValueError(f"{vector['id']}: verification differs from corpus")
        cases.append(
            {
                "caseId": vector["id"],
                "inputDigest": _digest(bytes(message)),
                "inputBytes": len(message),
                "parseOutcome": "accepted",
                "valid": verified,
                "findings": [] if verified else [
                    {"code": "SIGNATURE_INVALID", "instancePath": "/signature", "schemaPath": "/profile/ed25519"}
                ],
                "canonicalization": {"status": "not-applicable"},
                "componentDigest": None,
                "rootBomDigest": None,
                "signature": {"status": "verified"} if verified else {"status": "rejected", "code": "SIGNATURE_INVALID"},
            }
        )
    adapter_bytes = b"".join(
        (Path(__file__).parent / name).read_bytes()
        for name in ("signature.py", "signature_conformance.py")
    )
    return {
        "resultVersion": 1,
        "fixtureManifestDigest": _digest(corpus_bytes),
        "language": "python",
        "implementation": {
            "adapterId": "anvilkit-python-signature-native",
            "adapterVersion": "0.1.0",
            "runtime": "python",
            "runtimeVersion": "3.12.7",
            "adapterDigest": _digest(adapter_bytes),
        },
        "cases": cases,
    }


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--repository-root", type=Path, default=Path(__file__).resolve().parents[3])
    parser.add_argument("--iterations", type=int, default=1)
    args = parser.parse_args()
    if args.iterations < 1:
        parser.error("--iterations must be positive")
    result = None
    for _ in range(args.iterations):
        result = generate(args.repository_root)
    print(json.dumps(result, indent=2, ensure_ascii=False))


if __name__ == "__main__":
    main()
